vmm/src/memory_manager.rs

```diff
@@ -513,13 +518,17 @@ impl MemoryManager {
         Ok(Arc::new(match backing_file {
             Some(ref file) => {
                 let f = if file.is_dir() {
+                    // Override file offset as it does not apply in this case.
+                    file_offset = 0;
                     let fs_str = format!("{}{}", file.display(), "/tmpfile_XXXXXX");
                     let fs = ffi::CString::new(fs_str).unwrap();
                     let mut path = fs.as_bytes_with_nul().to_owned();
                     let path_ptr = path.as_mut_ptr() as *mut _;
                     let fd = unsafe { libc::mkstemp(path_ptr) };
                     unsafe { libc::unlink(path_ptr) };
-                    unsafe { File::from_raw_fd(fd) }
+                    let f = unsafe { File::from_raw_fd(fd) };
+                    f.set_len(size as u64).map_err(Error::SharedFileSetLen)?;
```

Comment:
I should split this into a dedicated commit. The point is that I think there was a bug as we were truncating the file even when it was a real file (and not a directory). There's no reason to truncate an existing file, as it can be mapped in two (or more) parts (basically what the PR is doing).